src/Microsoft.Identity.Client/TelemetryTokenCacheAccessor.cs

```diff
@@ -0,0 +1,92 @@
+﻿//----------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using Microsoft.Identity.Client.Internal;
+
+namespace Microsoft.Identity.Client
+{
+    internal class TelemetryTokenCacheAccessor : TokenCacheAccessor
+    {
+        // The content of this class has to be placed outside of its base class TokenCacheAccessor,
+        // otherwise we would have to modify multiple implementations of TokenCacheAccessor on different platforms.
+        public void SaveAccessToken(string cacheKey, string item, RequestContext requestContext)
+        {
+            var cacheEvent = new CacheEvent(CacheEvent.TokenCacheWrite) { TokenType = CacheEvent.TokenTypes.AT };
+            Telemetry.GetInstance().StartEvent(requestContext.TelemetryRequestId, cacheEvent);
```

Comment:
To be fair, I understand and appreciate @partnerinflight 's suggestion, and that's why I did try that approach at one point. It is a legitimate need for seeking a concise pattern to replace the boring try...finally... pattern.

And, as a fun fact, in Python there is a [context manager](https://docs.python.org/2/library/contextlib.html) concept and a `with` statement, which have same behavior as C#'s IDisposable and `using` statement. The only difference, AFAIK, is that in Python they don't discourage you from using such behavior to do whatever you see fit, while in C# they gave it a name to imply its use case scenario, and also explicitly say it in doc that "Implement IDisposable only if you are using unmanaged resources directly".

I've tried using `using`, it functionally work, and I personally loved it. But I foresaw it being controversal, so I did not take that move, otherwise this series of PRs would be able to be completed today.